other-cel/memory-requests-equal-limits/memory-requests-equal-limits.yaml

```diff
@@ -0,0 +1,38 @@
+apiVersion: kyverno.io/v1
+kind: ClusterPolicy
+metadata:
+  name: memory-requests-equal-limits
+  annotations:
+    policies.kyverno.io/title: Memory Requests Equal Limits in CEL expressions
+    policies.kyverno.io/category: Sample in CEL 
+    policies.kyverno.io/severity: medium
+    policies.kyverno.io/subject: Pod
+    policies.kyverno.io/minversion: 1.11.0
+    kyverno.io/kubernetes-version: "1.26-1.27"
+    policies.kyverno.io/description: >-
+      Pods which have memory limits equal to requests could be given a QoS class of Guaranteed if
+      they also set CPU limits equal to requests. Guaranteed is the highest schedulable class. 
+      This policy checks that all containers in a given Pod have memory requests equal to limits.
+spec:
+  validationFailureAction: Audit
+  background: false
+  rules:
+  - name: memory-requests-equal-limits
+    match:
+      any:
+      - resources:
+          kinds:
+          - Pod
+    validate:
+      cel:
+        variables:
+          - name: containersWithResources
+            expression: object.spec.containers.filter(container, has(container.resources))
+        expressions:
+        - expression: >-
+            variables.containersWithResources.all(container, 
+            !has(container.resources.requests) ||
+            !has(container.resources.requests.memory) ||
+            container.resources.requests.memory == container.resources.?limits.?memory.orValue('-1'))
```

Comment:
In this logic we check if `requests.memory` is equal to `limits.memory` or not.

We have 2 cases here:

1. `requests.memory` is not set and  `limits.memory` is either set or not set.

The logic should evaluate to true because `requests.memory` defaults to the value of `limits.memory` .
Here, is the description of `requests`.
```
$ k explain pod.spec.containers.resources.requests
KIND:       Pod
VERSION:    v1

FIELD: requests <map[string]Quantity>

DESCRIPTION:
    Requests describes the minimum amount of compute resources required. If
    Requests is omitted for a container, it defaults to Limits if that is
    explicitly specified, otherwise to an implementation-defined value. 
```

We use `!has(container.resources.requests) || !has(container.resources.requests.memory)` to account for this default behavior and return true.

2. `requests.memory` is set

i)  `limits.memory` is not set.

The logic should evaluate to false. And to do this, we have used the optional syntax `?` with `orValue()`.
So, in `container.resources.?limits.?memory.orValue('-1')`, if any of the fields `limits`, `memory` is not specified, the default value `-1` specified in `orValue() ` is compared with `requests.memory`, which will cause the logic to return a false, because `requests.memory` will surely not be the value -1 as it is invalid.
```
$ k apply -f -<<EOF
> apiVersion: v1
kind: Pod
metadata:
  name: badpod01
spec:
  containers:
  - name: busybox
    image: busybox:1.35
    resources:
      requests:
        memory: "-1"
> EOF
The Pod "badpod01" is invalid: spec.containers[0].resources.requests[memory]: Invalid value: "-1": must be greater than or equal to 0
```

ii) `limits.memory` is set.
In this case, the actual value of  `limits.memory` is directly compared with `requests.memory` and the `orValue('-1')` is not considered.

Not using the optional syntax can make the expression complex. 
Please let me know if it is clear.